centml/compiler/prediction/profiler.py

```diff
@@ -0,0 +1,226 @@
+import ast
+import csv
+import logging
+from typing import Dict
+
+import torch
+import torch.fx
+from sklearn.neighbors import KDTree  # type: ignore
+from torch.fx.node import Node
+
+from centml.compiler.config import settings
+
+
+class KDTreeWithValues:
+    def __init__(self, points=None, values=None):
+        self.points = points if points else []
+        self.values = values if values else []
+        if self.points:
+            self.tree = KDTree(self.points)
+        else:
+            self.tree = None
+
+    def add(self, point, value):
+        self.points.append(point)
+        self.values.append(value)
+        self.tree = KDTree(self.points)
+
+    def query(self, point):
+        if self.tree is None:
+            return None, None
+
+        dist, idx = self.tree.query([point], k=1)
+        return dist[0][0], self.values[idx[0][0]]
+
+
+class TreeDB:
+    def __init__(self):
+        self.db = {}
+
+    def add_from_db(self, key, points, times):
+        if key not in self.db:
+            self.db[key] = {}
+        self.db[key] = KDTreeWithValues(points, times)
+
+    def get(self, key, inp):
+        if key not in self.db:
+            logging.getLogger(__name__).warning(f"Key {key} not found in database")
+            return None
+
+        _, val = self.db[key].query(inp)
+        return val
+
+
+def populate_db(csv_file, database):
+    with open(csv_file, newline='') as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            try:
+                key = (row['op'], int(row['dim']), row['inp_dtypes'], row['out_dtypes'], row['gpu'])
+                points = ast.literal_eval(row['points'])
+                values = ast.literal_eval(row['values'])
+                database.add_from_db(key, points, values)
+            except ValueError as e:
+                logging.getLogger(__name__).exception(f"Error parsing row: {row}\n{e}")
+
+
+class Profiler:
+    def __init__(self, mod, gpu):
+        self.mod = mod
+        self.graph = mod.graph
+        self.modules = dict(self.mod.named_modules())
+        self.total_time = 0
+        self.TreeDB = TreeDB()
+        self.gpu = gpu
+        populate_db(settings.PREDICTION_DATA_FILE, self.TreeDB)
+
+    def propagate(self, *args):
+        args_iter = iter(args)
+        env: Dict[str, Node] = {}
+
+        dtypeMap = {
```

Comment:
yes. I had incorrectly assumed it would lead to duplicate entires (torch.float64 vs torch.double) but upon investigating they both print the same thing.